### ISSUE TITLE
PickList problem with too many POJOs

### ISSUE THREAD

**Issue by csaszika** [primefaces/primefaces#1739]
PickList works good with pojo in my project. But if I have too many pojos (in my case Action entities with Id, name, fk_actionType attributes) than picklList fails with this stacktrace:

`java.lang.NullPointerException
    at com.sun.faces.context.PartialViewContextImpl.createPartialResponseWriter(PartialViewContextImpl.java:469) [jsf-impl-2.1.28.redhat-3.jar:2.1.28.redhat-3]
    at com.sun.faces.context.PartialViewContextImpl.access$300(PartialViewContextImpl.java:76) [jsf-impl-2.1.28.redhat-3.jar:2.1.28.redhat-3]
    at com.sun.faces.context.PartialViewContextImpl$DelayedInitPartialResponseWriter.getWrapped(PartialViewContextImpl.java:603) [jsf-impl-2.1.28.redhat-3.jar:2.1.28.redhat-3]...`

Other informations: 
primefaces 6.0, 
JBoss 7.1.1.,
My backing bean looks like:
`package com.csaszi.productSheet;

import com.csaszi.abstractClasses.AbstractFormBean;
import com.csaszi.entities.*;
import com.csaszi.queryEnums.QueryEnumerations;
import com.csaszi.queryEnums.QueryParameterEnums;
import com.csaszi.remoteSessionBeans.QuerySessionBeanRemote;
import org.apache.log4j.Logger;
import org.primefaces.context.RequestContext;
import org.primefaces.event.CellEditEvent;
import org.primefaces.event.FlowEvent;
import org.primefaces.event.TransferEvent;
import org.primefaces.model.DualListModel;

import javax.annotation.PostConstruct;
import javax.ejb.EJB;
import javax.faces.application.FacesMessage;
import javax.faces.bean.ManagedBean;
import javax.faces.bean.ViewScoped;
import javax.faces.event.ActionEvent;
import javax.faces.event.ValueChangeEvent;
import java.util.*;

@ManagedBean
@ViewScoped
public class ProductSheetBuilderFormBean extends AbstractFormBean {

```
private static final Logger LOGGER = Logger.getLogger(ProductSheetBuilderFormBean.class);

@EJB
private QuerySessionBeanRemote<TCustomerEntity> queryCustomerSessionBean;
@EJB
private QuerySessionBeanRemote<TTemplateEntity> queryTemplateSessionBean;
@EJB
private QuerySessionBeanRemote<TActionEntity> queryActionSessionBean;
@EJB
private QuerySessionBeanRemote<TActionTypeEntity> queryActionTypeSessionBean;
@EJB
private QuerySessionBeanRemote<TempAction> queryTempActionSessionBean;
@EJB
private QuerySessionBeanRemote<TProductSheetNumEntity> queryProductSheetNumSessionBean;

/**
 * Header tab
 */
private Integer nextTlsz;   //to product sheet
private String product; //to product sheet
private Date startDate; //to product sheet
private Integer amount; //to product sheet
private TCustomerEntity customerToProductSheet;//to product sheet
private List<TCustomerEntity> customers; //for customer and filter for templates
/**
 * Template tab
 */
private List<TTemplateEntity> templates;
private TTemplateEntity chosenTemplate;
private TCustomerEntity chosenCustomerFilter;
/**
 * Operation codes tab
 */
private String newTemplateName;
//dialog
private List<TActionTypeEntity> actionTypes;
private TActionTypeEntity actionTypeEntity;
private String newActionName;
private DualListModel<TActionEntity> actions;
/**
 * Operation list with norms
 */
private int sumOfNorms;
private TempAction selectedTempActionEntity;
private List<TempAction> tempActions = new ArrayList<>();
//to product sheet and into database
private boolean templateSaved;

//SECOND
/**
 * Filter change on template tab
 */
public void onCustomerFilterChange(ValueChangeEvent event) {
    if (event.getNewValue() == null) {
        setTemplates(new ArrayList<TTemplateEntity>());
        setTempActions(new ArrayList<TempAction>());
        return;
    } else if (event.getNewValue() != getChosenCustomerFilter()){
        setChosenCustomerFilter((TCustomerEntity) event.getNewValue());
        setTempActions(new ArrayList<TempAction>());
    }

    EnumMap<QueryParameterEnums, Object> parameterMap = new EnumMap<>(QueryParameterEnums.class);
    parameterMap.put(QueryParameterEnums.CUSTOMER, event.getNewValue());
    setTemplates(queryTemplateSessionBean.queryReturnWithListWithParam(QueryEnumerations.TEMPLATES_FOR_CUSTOMER.getQuery(), parameterMap));
}

/**
 * Template change
 */
public void onTemplateChange(ValueChangeEvent event) {
    setChosenTemplate((TTemplateEntity) event.getNewValue());
    if (event.getNewValue() == null) {
        setTempActions(new ArrayList<TempAction>());
        fillUpActionsTab();
        return;
    }
    EnumMap<QueryParameterEnums, Object> parameterMap = new EnumMap<>(QueryParameterEnums.class);
    parameterMap.put(QueryParameterEnums.TEMPLATE, event.getNewValue());
    setTempActions(queryTempActionSessionBean.queryReturnWithListWithParam(QueryEnumerations.ACTIONS_AND_NORMS_BY_TEMPLATE.getQuery(), parameterMap));

    //refresh actions
    actions.getSource().clear();
    actions.setSource(queryActionSessionBean.queryReturnWithList(QueryEnumerations.ALL_ACTIONS_ORDER_BY_NAME.getQuery()));

    actions.getTarget().clear();
    for (TempAction res : tempActions) {
        actions.getTarget().add(res.getPk().getActionEntity());
    }
    actions.getSource().removeAll(actions.getTarget());

    growl(FacesMessage.SEVERITY_INFO, "Minta változás", "Műveletek sikeresen betöltve");
}

//THIRD
public void onTransfer(TransferEvent event) {
    StringBuilder builder = new StringBuilder();
    List<TActionEntity> actionEntities = new ArrayList<>((List<? extends TActionEntity>) event.getItems());
    //transfer direction (who get new actions)
    if (actions.getSource().contains(actionEntities.get(0))) {
        builder.append("Visszatétel: <br />");
        for (TActionEntity entity : actionEntities) {
            Iterator tempActionIterator = tempActions.iterator();
            while (tempActionIterator.hasNext()) {
                TempAction tA = (TempAction) tempActionIterator.next();
                if (tA.getActionEntity().equals(entity)) {
                    tempActionIterator.remove();
                }
            }
        }
    } else if (actions.getTarget().contains(actionEntities.get(0))){
        builder.append("Felvétel: <br />");
        for (TActionEntity entity : actionEntities) {
            TempAction tA = new TempAction();
            //set action, not persisted action hasn't got norm so 0
            tA.setNorm(0);
            tA.getPk().setActionEntity(entity);
            tempActions.add(tA);
        }
    } else {
        throw new IllegalArgumentException(actionEntities.get(0).getActionName()+
                    " element is not exist in source and in target side.");
    }

    paramContainer.setTempActions(getTempActions());

    //growl builder
    for(TActionEntity item : actionEntities) {
        builder.append(item.getActionName()).append("<br />");
    }

    growl(FacesMessage.SEVERITY_INFO, "Elem áthelyezés", builder.toString());
}

/**
 * Save action
 */
public void saveAction() {
    if (getNewActionName().equals("") || getActionTypeEntity() == null) {
        growl(FacesMessage.SEVERITY_WARN, "Hiba", "Töltsd ki az összes mezőt");
        return;
    }

    if (checkActionNameInDatabase()) {
        growl(FacesMessage.SEVERITY_WARN, "Hiba", "Már van művelet "+getActionTypeEntity().getActionTypeLetter()+"-"+getNewActionName()+" néven");
    } else {
        TActionEntity newActionEntity = new TActionEntity();
        newActionEntity.setActionName(getActionTypeEntity().getActionTypeLetter() + "-" + getNewActionName());

        newActionEntity.settActionTypeByActionTypeId(getActionTypeEntity());

        actionSessionBean.addEntity(newActionEntity);

        TempAction tA = new TempAction();
        //set action, not persisted action hasn't got norm so 0
        tA.setNorm(0);
        tA.getPk().setActionEntity((TActionEntity) queryActionSessionBean.queryReturnWithSingle(QueryEnumerations.ACTION_WITH_HIGHEST_ID.getQuery()));
        tempActions.add(tA);
        actions.getTarget().add(tA.getActionEntity());

        RequestContext.getCurrentInstance().execute("PF('newActionDialog').hide();");
        growl(FacesMessage.SEVERITY_INFO, "Új művelet", "Sikeres felvétel");
    }
}

private boolean checkActionNameInDatabase() {
    EnumMap<QueryParameterEnums, String> parameterMap = new EnumMap<>(QueryParameterEnums.class);
    parameterMap.put(QueryParameterEnums.ACTION_NAME, getActionTypeEntity().getActionTypeLetter() + "-" + getNewActionName());

    return querySessionBean.queryReturnSingleWithParam(QueryEnumerations.CHECK_ACTION_NAME.getQuery(), parameterMap) != null;
}

/**
 * Filter change on action tab's dialog
 */
public void onActionTypeFilterChange(ValueChangeEvent event) {
    setActionTypeEntity((TActionTypeEntity) event.getNewValue());
}

//Summary + LAST
private void orderingOperationList() {
    Collections.sort(tempActions, new Comparator<TempAction>() {
        public int compare(TempAction m1, TempAction m2) {
            return m1.getActionEntity().getActionName().compareTo(m2.getActionEntity().getActionName());
        }
    });
}

private void fixNormValuesOfOperations() {
    for (TempAction tA : getTempActions()) {
        if (tA.getNorm() == null) {
            tA.setNorm(0);
        }
    }
    RequestContext.getCurrentInstance().update("form:tempActionsTable");
}

public void onCellEdit(CellEditEvent event) {
    refreshSumOfNorm();
}

public void refreshSumOfNorm() {
    int sumOfNorms = 0;
    for (TempAction tA : getTempActions()){
        sumOfNorms+= tA.getNorm();
    }
    setSumOfNorms(sumOfNorms);
}

public void refreshTlszNum() {
    setNextTlsz((Integer) queryProductSheetNumSessionBean.queryReturnWithSingle(QueryEnumerations.PRODUCT_SHEET_NUM.getQuery()));
}

public void refreshMechanism(ActionEvent event) {
    refreshTlszNum();
    fillUpActionsTab();
    setTempActions(new ArrayList<TempAction>());
}

//OTHER
public String flowListener(FlowEvent event){
    System.out.println("Flow Event Happened :: New Step :: " + event.getNewStep() + " :: Old Step :: " + event.getOldStep());

    //CHECK HEADER DATA
    if (event.getOldStep().equals("header")) {
        //getAmount minimum is 1 so no check needs here
        if (getCustomerToProductSheet() == null || getProduct().equals("")) {
            growl(FacesMessage.SEVERITY_WARN,"Hiba", "Nincs kitöltve az összes mező");
            RequestContext.getCurrentInstance().update("form:growl");
            return event.getOldStep();
        }
    }
    //TO ACTIONS, NOT FROM SUMMARY
    if (event.getNewStep().equals("actions") &&
            !event.getOldStep().equals("summary") && getChosenTemplate() == null) {
        fillUpActionsTab();
    }
    //TO ACTIONS FROM SUMMARY
    if (event.getNewStep().equals("actions") &&
            event.getOldStep().equals("summary")){
        fillUpActionsTabOrderByTempActions();
        setTemplateSaved(false);
    }
    //GO TO SUMMARY
    if (event.getNewStep().equals("summary")) {
        fixNormValuesOfOperations();
        orderingOperationList();
        if (getTempActions().size() == 0) {
            growl(FacesMessage.SEVERITY_WARN, "Hiba", "Nincs hozzáadva művelet");
            growl(FacesMessage.SEVERITY_WARN, "Tipp", "Lehet, hogy visszaléptél a minta választáshoz, ilyenkor látsz elemeket a Használt oldalon. Változtasd meg a kiválasztott mintát másra, majd vissza az eredetire.");
            RequestContext.getCurrentInstance().update("form:growl");
            return event.getOldStep();
        }
    }
    //BACK FROM ACTIONS TO TEMPLATE CHOICE
    if (event.getNewStep().equals("templateChoice") &&
            event.getOldStep().equals("actions")) {
        setTempActions(new ArrayList<TempAction>());
        RequestContext.getCurrentInstance().update("form:tempActionsTable");
    }
    RequestContext.getCurrentInstance().update("form:toolbarButtonSet");
    return event.getNewStep();
}

//TEMPACTIONS TABLE
/**
 * Save template and filling of TempAction with contacts
 */
public void saveTemplate() {
    if (getNewTemplateName().equals("")) {
        growl(FacesMessage.SEVERITY_WARN, "Hiba", "Töltsd ki az összes mezőt");
        return;
    }

    if (checkTemplateNameInDatabase(getNewTemplateName())) {
        growl(FacesMessage.SEVERITY_WARN, "Hiba", "Már van mintád ezen a néven");
        setTemplateSaved(false);
    } else {
        TTemplateEntity newTemplateEntity = new TTemplateEntity();
        newTemplateEntity.setTemplateName(getNewTemplateName());
        newTemplateEntity.settCustomerByCustomerId(getCustomerToProductSheet());

        for (TempAction tA : getTempActions()) {
            tA.getPk().setTemplateEntity(newTemplateEntity);
        }
        newTemplateEntity.getTempActions().addAll(getTempActions());

        templateSessionBean.addEntity(newTemplateEntity);

        setTemplateSaved(true);
        RequestContext.getCurrentInstance().execute("PF('templateSaverDialog').hide();");

        growl(FacesMessage.SEVERITY_INFO, "Új minta", "Sikeres mentés");
    }
}

private boolean checkTemplateNameInDatabase(final String newName) {
    EnumMap<QueryParameterEnums, Object> parameterMap = new EnumMap<>(QueryParameterEnums.class);
    parameterMap.put(QueryParameterEnums.CUSTOMER, getCustomerToProductSheet());
    List<String> templateNames = new ArrayList<>(querySessionBean.queryReturnWithListBlockWithParam(QueryEnumerations.TEMPLATE_NAMES_FOR_CUSTOMER.getQuery(), parameterMap));

    for (String templateName : templateNames) {
        if (newName.equals(templateName)) {
            return true;
        }
    }
    return false;
}

//INIT
private void fillUpActionsTabOrderByTempActions() {
    List<TActionEntity> actionSource = new ArrayList<>(queryActionSessionBean.queryReturnWithList(QueryEnumerations.ALL_ACTIONS_ORDER_BY_NAME.getQuery()));
    List<TActionEntity> actionTarget = new ArrayList<>();

    for (TempAction res : tempActions) {
        actionTarget.add(res.getPk().getActionEntity());
    }
    actionSource.removeAll(actionTarget);

    actions = new DualListModel<>(actionSource, actionTarget);
}

public void onCustomerToProductSheetChange(ValueChangeEvent event) {
    setCustomerToProductSheet((TCustomerEntity) event.getNewValue());
}

private void fillUpActionsTab() {
    List<TActionEntity> actionsSource = new ArrayList<>(queryActionSessionBean.queryReturnWithList(QueryEnumerations.ALL_ACTIONS_ORDER_BY_NAME.getQuery()));
    List<TActionEntity> actionsTarget = new ArrayList<>();
    actions = new DualListModel<>(actionsSource, actionsTarget);
    //actionTypes in dialog
    actionTypes = new ArrayList<>(queryActionTypeSessionBean.queryReturnWithList(QueryEnumerations.ALL_ACTION_TYPES.getQuery()));
}

private void fillUpHeaderTab() {
    customers = new ArrayList<>(queryCustomerSessionBean.queryReturnWithList(QueryEnumerations.ALL_CUSTOMERS_ORDER_BY_NAME.getQuery()));
}

@PostConstruct
public void init() {
    fillUpHeaderTab();
    fillUpActionsTab();
    setNewTemplateName("");
    setProduct("");
    setAmount(1);
    setSumOfNorms(0);
    setNextTlsz((Integer) queryProductSheetNumSessionBean.queryReturnWithSingle(QueryEnumerations.PRODUCT_SHEET_NUM.getQuery()));
    Date today = Calendar.getInstance().getTime();
    setStartDate(today);
}

@Override
public String insert() {
    //functionless
    return null;
}

public boolean isTemplateSaved() {
    return templateSaved;
}

public void setTemplateSaved(boolean templateSaved) {
    this.templateSaved = templateSaved;
}

public TCustomerEntity getChosenCustomerFilter() {
    return chosenCustomerFilter;
}

public void setChosenCustomerFilter(TCustomerEntity chosenCustomerFilter) {
    this.chosenCustomerFilter = chosenCustomerFilter;
}

public TTemplateEntity getChosenTemplate() {
    return chosenTemplate;
}

public void setChosenTemplate(TTemplateEntity chosenTemplate) {
    this.chosenTemplate = chosenTemplate;
}

public TActionTypeEntity getActionTypeEntity() {
    return actionTypeEntity;
}

public void setActionTypeEntity(TActionTypeEntity actionTypeEntity) {
    this.actionTypeEntity = actionTypeEntity;
}

public String getNewActionName() {
    return newActionName;
}

public void setNewActionName(String newActionName) {
    this.newActionName = newActionName;
}

public List<TActionTypeEntity> getActionTypes() {
    return actionTypes;
}

public void setActionTypes(List<TActionTypeEntity> actionTypes) {
    this.actionTypes = actionTypes;
}

public int getSumOfNorms() {
    return sumOfNorms;
}

public void setSumOfNorms(int sumOfNorms) {
    this.sumOfNorms = sumOfNorms;
}

public List<TempAction> getTempActions() {
    return tempActions;
}

public void setTempActions(List<TempAction> tempActions) {
    this.tempActions = tempActions;
    paramContainer.setTempActions(tempActions);
}

public TempAction getSelectedTempActionEntity() {
    return selectedTempActionEntity;
}

public void setSelectedTempActionEntity(TempAction selectedTempActionEntity) {
    this.selectedTempActionEntity = selectedTempActionEntity;
}

public String getNewTemplateName() {
    return newTemplateName;
}

public void setNewTemplateName(String newTemplateName) {
    this.newTemplateName = newTemplateName;
}

public Integer getNextTlsz() {
    return nextTlsz;
}

public void setNextTlsz(Integer nextTlsz) {
    this.nextTlsz = nextTlsz;
    paramContainer.setTlsz(nextTlsz);
}

public TCustomerEntity getCustomerToProductSheet() {
    return customerToProductSheet;
}

public void setCustomerToProductSheet(TCustomerEntity customerToProductSheet) {
    this.customerToProductSheet = customerToProductSheet;
    paramContainer.setCustomerToProductSheet(customerToProductSheet);
}

public String getProduct() {
    return product;
}

public void setProduct(String product) {
    this.product = product;
    paramContainer.setProduct(product);
}

public Date getStartDate() {
    return startDate;
}

public void setStartDate(Date startDate) {
    this.startDate = startDate;
    paramContainer.setStartDate(startDate);
}

public Integer getAmount() {
    return amount;
}

public void setAmount(Integer amount) {
    this.amount = amount;
    paramContainer.setAmount(amount);
}

public List<TCustomerEntity> getCustomers() {
    return customers;
}

public void setCustomers(List<TCustomerEntity> customers) {
    this.customers = customers;
}

public List<TTemplateEntity> getTemplates() {
    return templates;
}

public void setTemplates(List<TTemplateEntity> templates) {
    this.templates = templates;
}

public DualListModel<TActionEntity> getActions() {
    return actions;
}

public void setActions(DualListModel<TActionEntity> actions) {
    this.actions = actions;
}
```

}
`
My ActionConverter for pickList looks like:
`@ManagedBean
@RequestScoped
@Named
public class ActionConverter extends AbstractEntityConverter implements Converter{

```
private static final Logger LOGGER = Logger.getLogger(ActionConverter.class);

@Override
public Object getAsObject(FacesContext context,
                          UIComponent component, String value) {
    if (value == null) {
        return null;
    }
    Integer id = new Integer(value);
    TActionEntity actionEntity = em.find(TActionEntity.class, id);
    if (actionEntity == null) {
        LOGGER.info("There is no actionEntity with id: "+id);
    }
    return actionEntity;
}

@Override
public String getAsString(FacesContext context,
                          UIComponent component, Object value) {
    if (value == null) {
        return "";
    }
    if (!(value instanceof TActionEntity)) {
        throw new IllegalArgumentException("This converter only handles instances of TActionEntity");
    }
    return ((TActionEntity)value).getActionId().toString();
}
```

}`
I know it is not FacesConverter, but this is one solution for entityManager.

I found two solutions:
`<context-param>
        <param-name>javax.faces.PARTIAL_STATE_SAVING</param-name>
        <param-value>true/false</param-value>
    </context-param>` - useless
PrimefacesExceptionHandler -useless
Yes there is more stupid workarounds(use just ID for itemValue and Name for ItemLabel in pickList), but I would like to use my pojos instead of playing with ID-s(always querying to my real entity).

My problematic pojo number is 524.
I think this can be a bug.

**Comment by tandraschko** [comment]
If you don't provide a small as possible example, i will close this as invalid.
Also the exception is not from us, it's from the JSF impl.

**Comment by csaszika** [comment]
The essence of the problem is that pickList has a limit number of pojos(entity), nothing more. If I use few entities there won't be any problem but if I use many entities in a pickList there will be this "exception". 
If you are sure that this problem doesn't belong to you than close it.
I made it on other way without pickList. DataTable can handle this 'big' number of pojo now.

**Comment by tandraschko** [comment]
We cant replicate such problems if you don't provide a small as possible example or test project.
Tested 5000 pojos in the showcase -> it's working fine.

**Comment by spadinha1** [comment]
Hello.. I am having the same problem... I made a filter to test.. and the problem is the delay to render picklist. 
Filtering on controller to 1000 pojos... its rendering ok... with all elements.. (8400) its not working.

**Comment by melloware** [comment]
@spadinha1 This was already fixed with this ticket for upcoming 6.3 release: https://github.com/primefaces/primefaces/issues/4320  It is also included in 6.2.14 if you are an Elite subscriber.
